File: src/main/java/komu/demodel/utils/IdentityHashSet.java
package komu.demodel.utils;

import java.util.AbstractSet;
import java.util.IdentityHashMap;
import java.util.Iterator;

public final class IdentityHashSet<T> extends AbstractSet<T> {

    private final IdentityHashMap<T, Object> map = new IdentityHashMap<T, Object>();
    
    @Override
    public Iterator<T> iterator() {
        return map.keySet().iterator();
    }

    @Override
    public int size() {
        return map.size();
    }
    
    @Override
    public boolean remove(Object o) {
        return map.remove(o) != null;
    }
    
    @Override
    public boolean add(T o) {
        return map.put(o, new Object()) == null;
    }
}

File: src/main/scala/komu/demodel/utils/IdentityHashSet.scala
/*
 * Copyright (C) 2010 Juha Komulainen. All rights reserved.
 */
package komu.demodel.utils

import java.util.{ AbstractSet, IdentityHashMap, Iterator }

final class IdentityHashSet[T] extends AbstractSet[T] {
  private val map = new IdentityHashMap[T,Object]
  
  def iterator = map.keySet.iterator  
  def size = map.size
  override def add(o: T) = map.put(o, new Object()) == null
  override def remove(o: Object) = map.remove(o) != null
}
